include/ginkgo/core/solver/multigrid.hpp

```diff
@@ -0,0 +1,417 @@
+
+
+enum class multigrid_mid_uses { pre, mid, post };
+/**
+ * Multigrid
+ *
+ * @tparam ValueType  precision of matrix elements
+ *
+ * @ingroup solvers
+ * @ingroup LinOp
+ */
+class Multigrid : public EnableLinOp<Multigrid> {
+    friend class EnableLinOp<Multigrid>;
+    friend class EnablePolymorphicObject<Multigrid, LinOp>;
+
+public:
+    /**
+     * Return true as iterative solvers use the data in x as an initial guess.
+     *
+     * @return true as iterative solvers use the data in x as an initial guess.
+     */
+    bool apply_uses_initial_guess() const override { return true; }
+
+    /**
+     * Gets the stopping criterion factory of the solver.
+     *
+     * @return the stopping criterion factory
+     */
+    std::shared_ptr<const stop::CriterionFactory> get_stop_criterion_factory()
+        const
+    {
+        return stop_criterion_factory_;
+    }
+
+    /**
+     * Sets the stopping criterion of the solver.
+     *
+     * @param other  the new stopping criterion factory
+     */
+    void set_stop_criterion_factory(
+        std::shared_ptr<const stop::CriterionFactory> other)
+    {
+        stop_criterion_factory_ = std::move(other);
+    }
+
+    /**
+     * Gets the system operator of the linear system.
+     *
+     * @return the system operator
+     */
+    std::shared_ptr<const LinOp> get_system_matrix() const
+    {
+        return system_matrix_;
+    }
+
+    /**
+     * Gets the list of MultigridLevel operator.
+     *
+     * @return the list of MultigridLevel opertor
+     */
+    const std::vector<std::shared_ptr<const gko::multigrid::MultigridLevel>>
+    get_mg_level_list() const
+    {
+        return mg_level_list_;
+    }
+
+    /**
+     * Gets the list of pre-smoother operator.
+     *
+     * @return the list of pre-smoother operator
+     */
+    const std::vector<std::shared_ptr<LinOp>> get_pre_smoother_list() const
+    {
+        return pre_smoother_list_;
+    }
+
+    /**
+     * Gets the list of mid-smoother operator.
+     *
+     * @return the list of mid-smoother operator
+     */
+    const std::vector<std::shared_ptr<LinOp>> get_mid_smoother_list() const
+    {
+        return mid_smoother_list_;
+    }
+
+    /**
+     * Gets the list of post-smoother operator.
+     *
+     * @return the list of post-smoother operator
+     */
+    const std::vector<std::shared_ptr<LinOp>> get_post_smoother_list() const
+    {
+        return post_smoother_list_;
+    }
+
+    /**
+     * Gets the operator at the coarsest level.
+     *
+     * @return the coarsest operator
+     */
+    std::shared_ptr<const LinOp> get_coarsest_solver() const
+    {
+        return coarsest_solver_;
+    }
+
+    /**
+     * Get the cycle of multigrid
+     *
+     * @return the cycle
+     */
+    multigrid_cycle get_cycle() const { return cycle_; }
+
+    /**
+     * Set the cycle of multigrid
+     *
+     * @param cycle the new cycle
+     */
+    void set_cycle(multigrid_cycle cycle) { cycle_ = cycle; }
```

Comment:
```suggestion
    void set_cycle(multigrid_cycle cycle) { parameters_.cycle_ = cycle; }
```

Reply:
okay, with this, the previous comments work.
I am not sure whether we have such of rules.
from others, we do not change the `parameters_` like criterion or preconditioner although they requires some generation after giving parameters.

Reply:
I just think it would be nice to have each parameter in only one variable in any given scope. Right now we have two different variables that should ideally hold the same information but in general they do not.